src/methods/scgpt_fine_tuned/functions.py

```diff
@@ -0,0 +1,529 @@
+                else ""
+                )
+            )
+            total_loss = 0
+            total_mse = 0
+            total_cls = 0
+            total_cce = 0
+            total_mvc = 0
+            total_ecs = 0
+            total_dab = 0
+            total_adv_E = 0
+            total_adv_D = 0
+            total_zero_log_prob = 0
+            total_mvc_zero_log_prob = 0
+            total_error = 0
+            start_time = time.time()
+
+
+def evaluate(
+    model,
+    loader,
+    device,
+    vocab,
+    pad_token,
+    hyperparameters,
+    training_settings,
+    criterion_cls,
+    criterion_dab,
+    return_raw
+):
+    """
+    Evaluate the model on the evaluation data.
+    """
+    model.eval()
+    total_loss = 0.0
+    total_error = 0.0
+    total_dab = 0.0
+    total_num = 0
+    predictions = []
+    with torch.no_grad():
+        for batch_data in loader:
+            input_gene_ids = batch_data["gene_ids"].to(device)
+            input_values = batch_data["values"].to(device)
+            target_values = batch_data["target_values"].to(device)
+            batch_labels = batch_data["batch_labels"].to(device)
+            if not return_raw:
+                celltype_labels = batch_data["celltype_labels"].to(device)
+
+            src_key_padding_mask = input_gene_ids.eq(vocab[pad_token])
+            with torch.cuda.amp.autocast(enabled=hyperparameters["amp"]):
+                output_dict = model(
+                    input_gene_ids,
+                    input_values,
+                    src_key_padding_mask=src_key_padding_mask,
+                    batch_labels=batch_labels if hyperparameters["DSBN"] else None,
+                    CLS=training_settings["CLS"],  # evaluation does not need CLS or CCE
+                    CCE=False,
+                    MVC=False,
+                    ECS=False,
+                    do_sample=False,
+                )
+                output_values = output_dict["cls_output"]
+                print(output_values)
+                if not return_raw:
+                    loss = criterion_cls(output_values, celltype_labels)
+                    if training_settings["DAB"]:
+                        loss_dab = criterion_dab(output_dict["dab_output"], batch_labels)
+        
+            if return_raw:
+                preds = output_values.argmax(1).cpu().numpy()
+                predictions.append(preds)
+            else:
+                total_loss += loss.item() * len(input_gene_ids)
+                accuracy = (output_values.argmax(1) == celltype_labels).sum().item()
+                total_error += (1 - accuracy / len(input_gene_ids)) * len(input_gene_ids)
+                total_dab += loss_dab.item() * len(input_gene_ids) if training_settings["DAB"] else 0.0
+                total_num += len(input_gene_ids)
+
+    if return_raw:
+      return np.concatenate(predictions, axis=0)
+
+    return total_loss / total_num, total_error / total_num
+
+
+def test(
```

Comment:
I don't think this function is used anywhere so maybe it could be removed?

Reply:
This function is used to get the predictions.